pythia/utils/text_utils.py

```diff
@@ -243,53 +296,113 @@ def search(self, t, data, scores):
-            self.complete_seqs.extend(self.seqs[complete_inds].tolist())
-            self.complete_seqs_scores.extend(self.top_k_scores[complete_inds])
+            self._complete_seqs.extend(self.seqs[complete_inds].tolist())
+            self._complete_seqs_scores.extend(self.top_k_scores[complete_inds])
 
         # Reduce beam length
-        self.beam_size -= len(complete_inds)
+        self._decode_size -= len(complete_inds)
 
         # Proceed with incomplete sequences
-        if self.beam_size == 0:
+        if self._decode_size == 0:
             return True, data, 0
 
         self.seqs = self.seqs[incomplete_inds]
         self.top_k_scores = self.top_k_scores[incomplete_inds].unsqueeze(1)
 
         # TODO: Make the data update generic for any type of model
         # This is specific to BUTD model only.
-        data["texts"] = next_word_inds[incomplete_inds].unsqueeze(1)
-        h1 = data["state"]["td_hidden"][0][prev_word_inds[incomplete_inds]]
-        c1 = data["state"]["td_hidden"][1][prev_word_inds[incomplete_inds]]
-        h2 = data["state"]["lm_hidden"][0][prev_word_inds[incomplete_inds]]
-        c2 = data["state"]["lm_hidden"][1][prev_word_inds[incomplete_inds]]
-        data["state"] = {"td_hidden": (h1, c1), "lm_hidden": (h2, c2)}
+        data = self.update_data(data, prev_word_inds, next_word_inds, incomplete_inds)
 
         next_beam_length = len(prev_word_inds[incomplete_inds])
 
         return False, data, next_beam_length
 
-    def best_score(self):
-        if len(self.complete_seqs_scores) == 0:
+    def get_result(self):
+        if len(self._complete_seqs_scores) == 0:
             captions = torch.FloatTensor([0] * 5).unsqueeze(0)
         else:
-            i = self.complete_seqs_scores.index(max(self.complete_seqs_scores))
-            captions = torch.FloatTensor(self.complete_seqs[i]).unsqueeze(0)
+            i = self._complete_seqs_scores.index(max(self._complete_seqs_scores))
+            captions = torch.FloatTensor(self._complete_seqs[i]).unsqueeze(0)
+        return captions
+
+
+class NucleusSampling(TextDecoder):
+    """Nucleus Sampling is a new text decoding strategy that avoids likelihood maximization.
+    Rather, it works by sampling from the smallest set of top tokens which have a cumulative
+    probability greater than a specified threshold.
+
+    Present text decoding strategies like beam search do not work well on open-ended
+    generation tasks (even on strong language models like GPT-2). They tend to repeat text
+    a lot and the main reason behind it is that they try to maximize likelihood, which is a
+    contrast from human-generated text which has a mix of high and low probability tokens.
+
+    Nucleus Sampling is a stochastic approach and resolves this issue. Moreover, it improves
+    upon other stochastic methods like top-k sampling by choosing the right amount of tokens
+    to sample from. The overall result is better text generation on the same language model.
+
```

Comment:
Can you also add a link to the paper (with section number) so that people can refer to the paper with more details